compiler/insert/type_check.rs

```diff
@@ -11,28 +11,20 @@ use std::{
 
 use answer::{variable::Variable, Type};
 use concept::type_::type_manager::TypeManager;
-use ir::{pattern::constraint::Constraint, program::block::FunctionalBlock};
+use ir::{
+    pattern::constraint::{Constraint, Has, Links},
+    program::block::FunctionalBlock,
+};
+use itertools::{chain, Itertools};
 use storage::snapshot::ReadableSnapshot;
 
 use crate::match_::inference::{
-    type_annotations::{ConstraintTypeAnnotations, TypeAnnotations},
+    type_annotations::{
+        ConstraintTypeAnnotations, LeftRightAnnotations, LeftRightFilteredAnnotations, TypeAnnotations,
+    },
     TypeInferenceError,
 };
 
-pub enum ValidCombinations<'a> {
-    Has(&'a BTreeMap<answer::Type, Vec<answer::Type>>),
-    Links(&'a BTreeMap<Type, HashSet<Type>>),
-}
-
-impl<'a> ValidCombinations<'a> {
-    fn check(&self, left: &answer::Type, right: &answer::Type) -> bool {
-        match self {
-            ValidCombinations::Has(has) => has.get(left).unwrap().contains(right),
-            ValidCombinations::Links(links) => links.get(left).unwrap().contains(right),
-        }
-    }
-}
-
```

Comment:
I would have needed this data-structure for a more fine-grained check, but we'll just reinvent it when we get to it.